tianshou/data/batch.py

```diff
@@ -537,33 +545,42 @@ def stack(batches: List['Batch'], axis: int = 0) -> 'Batch':
         batch.stack_(batches, axis)
         return batch
 
-    def empty_(self) -> 'Batch':
+    def empty_(self, index: Union[
+        str, slice, int, np.integer, np.ndarray, List[int]] = None
+    ) -> 'Batch':
         """Return an empty a :class:`~tianshou.data.Batch` object with 0 or
-        ``None`` filled.
+        ``None`` filled. If ``index`` is specified, it will only reset the
+        specific indexed-data.
         """
         for k, v in self.items():
             if v is None:
                 continue
             if isinstance(v, Batch):
-                self.__dict__[k].empty_()
-            elif isinstance(v, np.ndarray) and v.dtype == np.object:
-                self.__dict__[k].fill(None)
-            elif isinstance(v, torch.Tensor):  # cannot apply fill_ directly
-                self.__dict__[k] = torch.zeros_like(self.__dict__[k])
-            else:  # np
-                self.__dict__[k] *= 0
-                if hasattr(v, 'dtype') and v.dtype.kind in 'fc':
-                    self.__dict__[k] = np.nan_to_num(self.__dict__[k])
+                self.__dict__[k].empty_(index=index)
+            elif isinstance(v, torch.Tensor):
+                self.__dict__[k][index] = 0
+            elif isinstance(v, np.ndarray):
+                if v.dtype == np.object:
+                    self.__dict__[k][index] = None
+                else:
+                    self.__dict__[k][index] = 0
+            else:  # scalar value
+                if isinstance(v, (np.generic, Number)):
+                    self.__dict__[k] = 0
```

Comment:
It does not preserve `dtype`, but `self.__dict__[k] *= 0` does.

Reply:
But *=0 will not resolve the nan issue:
```python
b = nan
b *= 0
# b is still nan
```

Reply:
Yes but it is an issue ?